kvserver: tease apart applySnapshot

tbg opened the pull request:
TODO

Epic: CRDB-46488
Release note: none

tbg committed:
kvserver: add snapshot_prepare.go

We'll extract all the engine interactions from `(*Replica).applySnapshot` into this file.

The intention is that after that point, we will extract into a package and add
unit testing.

Epic: CRDB-46488
Release note: none

tbg committed:
kvserver: move subsumedDescs computation

This makes it easier to extract storage-related code in follow-up commits.

Epic: CRDB-46488
Release note: none

tbg committed:
kvserver: applySnapshot extraction, step 1

Create a stub `prepareSnapshot` method which will be a free-standing function
doing all of the engine work.

For now, it only calls `writeUnreplicatedSST`. We'll be adding more in follow-up
commits.

Epic: CRDB-46488
Release note: none

tbg committed:
kvserver: applySnapshot extraction, step 2

Move clearSubsumedReplicaDiskData to prepareSnapshot.

Epic: CRDB-46488
Release note: none

tbg committed:
kvserver: applySnapshot extraction, step 3

Move a call to writeSST into prepareSnapshot.

Epic: CRDB-46488
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142431)
<!-- Reviewable:end -->

tbg committed:
kvserverpb: add migration note on RangeKeysInOrder bool

tbg committed:
kvserver: simplify snap ingestion logic

There was duplicate code in two branches. Streamline the code to make it clear
that there are three cases:

- excise-based ingest
- clearrange-based ingest (SSTs contain the deletions)
- ingest as write batch (for small snaps).

This will be teased apart further in future commits.

Epic: CRDB-46488
Release note: none

tbg commented:
Superseded by https://github.com/cockroachdb/cockroach/pull/145328.
